File: common/src/main/java/juuxel/adorn/block/variant/BlockVariantSet.java
package juuxel.adorn.block.variant;

import java.util.List;

public interface BlockVariantSet {
    default List<BlockVariant> getWoodVariants() {
        return List.of();
    }

    default List<BlockVariant> getStoneVariants() {
        return List.of();
    }

    default void addVariants(CustomVariantConsumer consumer) {
        // No customised variants by default.
    }

    default void sortVariants(Sorter sorter) {
        // No custom sorting by default.
    }

    @FunctionalInterface
    interface CustomVariantConsumer {
        void add(BlockVariant variant, List<BlockKind> kinds);
    }

    @FunctionalInterface
    interface Sorter {
        void moveAfter(BlockVariant variant, BlockVariant after);
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/variant/BlockVariantSet.kt
package juuxel.adorn.block.variant

import java.util.function.Function

interface BlockVariantSet {
    val woodVariants: List<BlockVariant> get() = emptyList()
    val stoneVariants: List<BlockVariant> get() = emptyList()

    fun addVariants(consumer: CustomVariantConsumer) {
        // No customised variants by default.
    }

    fun sortVariants(sorter: Sorter) {
        // No custom sorting by default.
    }

    companion object {
        @JvmStatic
        fun createVariants(prefix: String, factory: Function<String, BlockVariant>, vararg variants: String): List<BlockVariant> =
            variants.map { factory.apply("$prefix/$it") }
    }

    fun interface CustomVariantConsumer {
        fun add(variant: BlockVariant, kinds: List<BlockKind>)
    }

    fun interface Sorter {
        fun moveAfter(variant: BlockVariant, after: BlockVariant)
    }
}
